build.assets/tooling/lib/tree/tree.go

```diff
@@ -0,0 +1,334 @@
+
+	"github.com/gogo/protobuf/gogoproto"
+	"github.com/gogo/protobuf/proto"
+	gogodesc "github.com/gogo/protobuf/protoc-gen-gogo/descriptor"
+	gogogen "github.com/gogo/protobuf/protoc-gen-gogo/generator"
+)
+
+// Forest is a forest of file trees. It's a wrapper for generator.Generator.
+type Forest struct {
+	*gogogen.Generator
+	messageMap map[*gogodesc.DescriptorProto]*Message
+}
+
+func NewForest(gen *gogogen.Generator) *Forest {
+	return &Forest{
+		Generator:  gen,
+		messageMap: make(map[*gogodesc.DescriptorProto]*Message),
+	}
+}
+
+// File is a wrapper for generator.FileDescriptor
+type File struct {
+	parent        *Forest
+	desc          *gogodesc.FileDescriptorProto
+	locations     map[string]Location
+	messageMap    map[*gogodesc.DescriptorProto]*Message
+	messageByName map[string]*Message
+
+	Messages []*Message
+}
+
+// Message is a wrapper for generator.Descriptor.
+type Message struct {
+	index     int
+	parent    *File
+	parentMsg *Message
+	desc      *gogodesc.DescriptorProto
+	Fields    []*Field
+	fieldMap  map[string]*Field
+}
+
+// Field is wrapper for descriptor.FieldDescriptorProto.
+type Field struct {
+	index  int
+	parent *Message
+	desc   *gogodesc.FieldDescriptorProto
+}
+
+// Location is a wrapper for descriptor.SourceCodeInfo_Location.
+type Location struct {
+	loc *gogodesc.SourceCodeInfo_Location
+}
+
+func (forest *Forest) AddFile(fileDesc *gogodesc.FileDescriptorProto) *File {
+	locs := fileDesc.GetSourceCodeInfo().GetLocation()
+	msgs := fileDesc.GetMessageType()
+	file := File{
+		parent:        forest,
+		desc:          fileDesc,
+		locations:     make(map[string]Location, len(locs)),
+		Messages:      make([]*Message, len(msgs)),
+		messageMap:    make(map[*gogodesc.DescriptorProto]*Message, len(msgs)),
+		messageByName: make(map[string]*Message, len(msgs)),
+	}
+
+	// Build locations map.
+	for _, loc := range locs {
+		path := loc.GetPath()
+		pathStr := make([]string, len(path))
+		for i, point := range path {
+			pathStr[i] = strconv.Itoa(int(point))
+		}
+		file.locations[strings.Join(pathStr, ",")] = Location{loc: loc}
+	}
+
+	// Build messages list.
+	for i, msgDesc := range msgs {
+		message := forest.addMessageFromDesc(msgDesc, i, &file, nil)
+		file.Messages[i] = message
+		file.messageMap[msgDesc] = message
+		file.messageByName[message.Name()] = message
+	}
+
+	return &file
+}
+
+// addMessageFromDesc generates a message from a gogodesc.DescriptorProto and registers the message into the forest messageMap.
+// This function will recursively register all nested messages if any, but only returns the root message.
+func (forest *Forest) addMessageFromDesc(msgDesc *gogodesc.DescriptorProto, index int, file *File, parentMsg *Message) *Message {
+	msgDesc.GetName()
+	flds := msgDesc.GetField()
+	message := Message{
+		index:     index,
+		desc:      msgDesc,
+		parent:    file,
+		parentMsg: parentMsg,
+		Fields:    make([]*Field, 0, len(flds)),
+		fieldMap:  make(map[string]*Field, len(flds)),
```

Comment:
I'm curios about ordering. Namely map doesn't guarantee that elements order will be preserved. Not sure how we will implement view update logic but I think that different field order should not affect view update logic.

Reply:
I changed the `Event.TableSchema()` and `Event.ViewSchema()` functions to sort fields before iterating over them.